crates/hir/src/doc_links.rs

```diff
@@ -0,0 +1,226 @@
+//! Resolves links in markdown documentation.
+
+use std::iter::once;
+
+use itertools::Itertools;
+use url::Url;
+
+use crate::{db::HirDatabase, Adt, AsName, Crate, Hygiene, ItemInNs, ModPath, ModuleDef};
+use hir_def::{db::DefDatabase, resolver::Resolver};
+use syntax::ast::Path;
+
+pub fn resolve_doc_link<T: Resolvable + Clone, D: DefDatabase + HirDatabase>(
+    db: &D,
+    definition: &T,
+    link_text: &str,
+    link_target: &str,
+) -> Option<(String, String)> {
+    try_resolve_intra(db, definition, link_text, &link_target).or_else(|| {
+        if let Some(definition) = definition.clone().try_into_module_def() {
+            try_resolve_path(db, &definition, &link_target)
+                .map(|target| (target, link_text.to_string()))
+        } else {
+            None
+        }
+    })
+}
+
+/// Try to resolve path to local documentation via intra-doc-links (i.e. `super::gateway::Shard`).
+///
+/// See [RFC1946](https://github.com/rust-lang/rfcs/blob/master/text/1946-intra-rustdoc-links.md).
+fn try_resolve_intra<T: Resolvable, D: DefDatabase + HirDatabase>(
```

Comment:
It would be better if we can make this on-generic:

https://github.com/rust-analyzer/rust-analyzer/pull/5853/files

Reply:
Agreed, but I'm not sure it's easy/possible with the `DefDatabase + HirDatabase` bound (which also exists in `Resolvable`).

I know you can declare a trait:

```rust
trait DB: DefDatabase + HirDatabase {}
```

And then create `&dyn DB` objects, but I don't think there's an easy way to get `&dyn DefDatabase`/`&dyn HirDatabase` out of those?